core/src/test/java/com/linecorp/armeria/server/ServerMaxConnectionAgeTest.java

```diff
@@ -0,0 +1,161 @@
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.testing.junit.server.ServerExtension;
+
+import io.netty.util.AttributeMap;
+
+class ServerMaxConnectionAgeTest {
+
+    private static final int MAX_CONNECTION_AGE_MILLIS = 5000;
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.idleTimeoutMillis(0);
+            sb.maxConnectionAgeMillis(MAX_CONNECTION_AGE_MILLIS);
+            sb.service("/", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
+        }
+    };
+
+    @RegisterExtension
+    static ServerExtension serverKeepAlive = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.idleTimeoutMillis(0);
+            sb.service("/", (ctx, req) -> HttpResponse.of(HttpStatus.OK));
+        }
+    };
+
+    AtomicInteger opened;
+    AtomicInteger closed;
+    ConnectionPoolListener connectionPoolListener;
+
+    @BeforeEach
+    void setUp() {
+        opened = new AtomicInteger();
+        closed = new AtomicInteger();
+        connectionPoolListener = new ConnectionPoolListener() {
+            @Override
+            public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                       InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                opened.incrementAndGet();
+            }
+
+            @Override
+            public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                         InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                closed.incrementAndGet();
+            }
+        };
+    }
+
+    @CsvSource({ "H1C, false", "H1C, true", "H2C, true" })
+    @ParameterizedTest
+    void shouldDisconnectWhenMaxConnectionAgeIsExceeded(SessionProtocol protocol, boolean useHttp1Pipelining)
+            throws InterruptedException {
+        final ClientFactory clientFactory =
+                ClientFactory.builder()
+                             .connectionPoolListener(connectionPoolListener)
+                             .useHttp1Pipelining(useHttp1Pipelining)
+                             .idleTimeoutMillis(0)
+                             .build();
+        final WebClient client =
+                WebClient.builder(server.uri(protocol))
+                         .factory(clientFactory)
+                         .build();
+
+        final Stopwatch stopwatch = Stopwatch.createStarted();
+        final AtomicInteger oldClosed = new AtomicInteger();
+
+        while (closed.get() < 5) {
+            assertThat(client.get("/").aggregate().join().status()).isEqualTo(HttpStatus.OK);
```

Comment:
I guess we don't really test the case of HTTP/1 pipelining enabled because we always wait until the response is received? Could we have dedicated tests for:

- HTTP/1 with pipelining enabled, so that the requests handled after the server sends `Connection: close` fails with some exception, e.g. `ClosedSessionException`, making sure they don't go through the decorator chain at all.
- HTTP/2 whth simultaneous request, so that we make sure the requests whose stream ID is greater than `lastStreamId` fail with an `UnprocessedRequestException(GoAwayReceivedException)`, making sure they don't through the decorator chain at all.

Reply:
> I guess we don't really test the case of HTTP/1 pipelining enabled because we always wait until the response is received?

Ahh... you are right. Let me improve the test cases as you suggested.